src/core/sock/sockinfo_tcp.cpp

```diff
@@ -6042,3 +6029,192 @@ inline bool sockinfo_tcp::handle_bind_no_port(int &bind_ret, in_port_t in_port,
 
     return CONTINUE_WITH_BIND;
 }
+
+int sockinfo_tcp::tcp_tx_express(const struct iovec *iov, unsigned iov_len, uint32_t mkey,
+                                 xlio_express_flags flags, void *opaque_op)
+{
+    if (unlikely(!is_rts())) {
+        if (m_conn_state == TCP_CONN_TIMEOUT) {
+            si_tcp_logdbg("TX timed out");
+            errno = ETIMEDOUT;
+        } else if (m_conn_state == TCP_CONN_RESETED) {
+            si_tcp_logdbg("TX on reseted socket");
+            errno = ECONNRESET;
+        } else if (m_conn_state == TCP_CONN_ERROR) {
+            si_tcp_logdbg("TX on connection failed socket");
+            errno = ECONNREFUSED;
+        } else {
+            si_tcp_logdbg("TX on disconnected socket");
+            errno = EPIPE;
+        }
+        return -1;
+    }
+
+    err_t err;
+    pbuf_desc mdesc;
+
+    switch (flags & XLIO_EXPRESS_OP_TYPE_MASK) {
+    case XLIO_EXPRESS_OP_TYPE_DESC:
+        mdesc.attr = PBUF_DESC_EXPRESS;
+        break;
+    case XLIO_EXPRESS_OP_TYPE_FILE_ZEROCOPY:
+        mdesc.attr = PBUF_DESC_MDESC;
+        break;
+    default:
+        return -1;
+    };
+    mdesc.mkey = mkey;
+    mdesc.opaque = opaque_op;
+
+    int bytes_written = 0;
+
+    lock_tcp_con();
+
+    for (unsigned i = 0; i < iov_len; ++i) {
+        err = tcp_write_express(&m_pcb, iov[i].iov_base, iov[i].iov_len, &mdesc);
+        if (err != ERR_OK) {
+            return -1;
```

Comment:
We need either:
1. Move socket to an error state, so user may not work with it
2. Or revert all the tcp_write_express() segments, so we don't corrupt TCP stream with send() retry.

Reply:
I think for this stage option 1 is good.

Reply:
Done